src/rgw/driver/d4n/d4n_datacache.cc

```diff
@@ -0,0 +1,482 @@
+#include "d4n_datacache.h"
+
+#define dout_subsys ceph_subsys_rgw
+#define dout_context g_ceph_context
+
+/* Base metadata and data fields should remain consistent */
+std::vector<std::string> baseFields{
+  "mtime",
+  "object_size",
+  "accounted_size",
+  "epoch",
+  "version_id",
+  "source_zone_short_id",
+  "bucket_count",
+  "bucket_size",
+  "user_quota.max_size",
+  "user_quota.max_objects",
+  "max_buckets",
+  "data"};
+
+std::vector< std::pair<std::string, std::string> > RGWD4NCache::buildObject(rgw::sal::Attrs* binary) {
+  std::vector< std::pair<std::string, std::string> > values;
+  rgw::sal::Attrs::iterator attrs;
+ 
+  /* Convert to vector */
+  if (binary != NULL) {
+    for (attrs = binary->begin(); attrs != binary->end(); ++attrs) {
+      values.push_back(std::make_pair(attrs->first, attrs->second.to_str()));
+    }
+  } 
+
+  return values; 
+}
+
+int RGWD4NCache::findClient(cpp_redis::client *client) { 
+  if (client->is_connected())
+    return 0;
+
+  if (host == "" || port == 0) { 
+    dout(10) << "RGW D4N Cache: D4N cache endpoint was not configured correctly" << dendl;
+    return EDESTADDRREQ;
+  }
+
+  client->connect(host, port, nullptr);
+
+  if (!client->is_connected())
+    return ECONNREFUSED;
+
+  return 0;
+}
+
+int RGWD4NCache::existKey(std::string key) { 
+  int result = -1;
+  std::vector<std::string> keys;
+  keys.push_back(key);
+
+  if (!client.is_connected()) {
+    return result;
+  }
+
+  try {
+    client.exists(keys, [&result](cpp_redis::reply &reply) {
+      if (reply.is_integer()) {
+        result = reply.as_integer();
+      }
+    });
+
+    client.sync_commit(std::chrono::milliseconds(1000));
+  } catch(std::exception &e) {}
+  
+  return result;
+}
+
+int RGWD4NCache::setObject(std::string oid, rgw::sal::Attrs* attrs) {
+  /* Creating the index based on oid */
+  std::string key = "rgw-object:" + oid + ":cache";
+  std::string result;
+
+  if (!client.is_connected()) {
+    findClient(&client);
+  }
+
+  /* Every set will be treated as new */
+  try {
+    std::vector< std::pair<std::string, std::string> > redisObject = buildObject(attrs);
+      
+    if (redisObject.empty()) {
+      return -1;
+    }
+      
+    client.hmset(key, redisObject, [&result](cpp_redis::reply &reply) {
+      if (!reply.is_null()) {
+        result = reply.as_string();
+      }
+    });
+
+    client.sync_commit(std::chrono::milliseconds(1000));
+
+    if (result != "OK") {
+      return -1;
+    }
+  } catch(std::exception &e) {
+    return -2;
```

Comment:
@samarahu I'm curious what's the point of returning `-2` here and some of the other negative values that are not `-1` in the functions below.

Reply:
I'll have to check if it's consistent but -2 is to indicate the key doesn't exist while -1 is for any other errors. This will be useful in the future when determining where objects/blocks are located.